src/main/java/io/camunda/testing/assertions/ProcessInstanceAssertions.java

```diff
@@ -0,0 +1,413 @@
+  }
+
+  /**
+   * Verifies the expectation that the process instance is currently not waiting at one or more
+   * specified elements.
+   *
+   * @param elementIds The ids of the elements
+   * @return this {@link ProcessInstanceAssertions}
+   */
+  public ProcessInstanceAssertions isNotWaitingAtElement(final String... elementIds) {
+    final List<String> activatedElements =
+        Arrays.stream(elementIds).filter(this::isWaitingAtElement).collect(Collectors.toList());
+
+    assertThat(activatedElements.isEmpty())
+        .withFailMessage(
+            "Process with key %s is waiting at element(s) with id(s) %s",
+            actual.getProcessInstanceKey(), String.join(", ", activatedElements))
+        .isTrue();
+
+    return this;
+  }
+
+  /**
+   * Checks if the process instance is currently waiting at an element with a specific element id.
+   *
+   * @param elementId The id of the element
+   * @return boolean indicating whether the process instance is waiting at the element
+   */
+  // TODO if waiting at any element with this id returns true. Maybe add a counter so we can check
+  // we are waiting at multiple elements with the same id
+  private boolean isWaitingAtElement(final String elementId) {
+    final Map<Long, List<Record<ProcessInstanceRecordValue>>> recordsMap =
+        StreamFilter.processInstance(recordStreamSource.processInstanceRecords())
+            .withProcessInstanceKey(actual.getProcessInstanceKey()).withElementId(elementId)
+            .stream()
+            .collect(Collectors.groupingBy(record -> record.getValue().getFlowScopeKey()));
+
+    if (recordsMap.isEmpty()) {
+      return false;
+    }
+
+    for (Entry<Long, List<Record<ProcessInstanceRecordValue>>> entry : recordsMap.entrySet()) {
+      final Record<ProcessInstanceRecordValue> lastRecord =
+          entry.getValue().get(entry.getValue().size() - 1);
+      if (lastRecord.getIntent() == ProcessInstanceIntent.ELEMENT_ACTIVATED) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  /**
+   * Verifies the expectation that the process instance is currently waiting at the specified
+   * elements, and not at any other element.
+   *
+   * @param elementIds The ids of the elements
+   * @return this {@link ProcessInstanceAssertions}
+   */
+  public ProcessInstanceAssertions isWaitingExactlyAtElements(final String... elementIds) {
+    final List<String> shouldBeWaitingAt = Arrays.asList(elementIds);
+    final List<String> wrongfullyWaitingElementIds = new ArrayList<>();
+    final List<String> wrongfullyNotWaitingElementIds = new ArrayList<>();
+
+    StreamFilter.processInstance(recordStreamSource.processInstanceRecords())
+        .withProcessInstanceKey(actual.getProcessInstanceKey())
+        .withoutBpmnElementType(BpmnElementType.PROCESS)
+        .withIntent(ProcessInstanceIntent.ELEMENT_ACTIVATING).stream()
+        .map(Record::getValue)
+        .map(ProcessInstanceRecordValue::getElementId)
+        .forEach(
```

Comment:
:x: I think this could be refactored to avoid unnecessary calculation.

* the map of element IDs could be reduced to its unique elements
* `isWaitingAtElement(id)` could be cached
* `shouldBeWaitingAt.contains(id)` could be cached

I think this would also make the logic clearer.